### PR TITLE
Add operation data into Argo Application

### PR DESCRIPTION
**Pull request by JohnNiang**
### What type of PR is this?

/kind feature
/kind cleanup

### What this PR does / why we need it:

As I mentioned in the title, I added operation data into Argo Application and move argoApp field downward.

### Which issue(s) this PR fixes:

Fixes #

### Special notes for reviewers:

Please check the following list before waiting reviewers:

- [ ] Already committed the CRD files to [the Helm Chart](https://github.com/kubesphere-sigs/ks-devops-helm-chart/) if you created some new CRDs
- [ ] Already [added the permission](https://github.com/kubesphere/ks-installer/blob/9e063b085a0e43fdb3d0d9e3e7f4149146f14b9c/roles/ks-core/prepare/files/ks-init/role-templates.yaml) for the new API
- [ ] Already added the RBAC markers for the new controllers

### Does this PR introduce a user-facing change??

```release-note
None
```

/cc @kubesphere/sig-devops

### CODE REVIEW

File: controllers/argocd/utils.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+package argocd
+
+import (
+	"encoding/json"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// StructToMap converts struct value to map value, which key type is string and value type is interface{}.
+// The type of value must be struct. Any other types will lead an error.
+func StructToMap(value interface{}) (map[string]interface{}, error) {
```
Comment (LinuxSuRen) [inline controllers/argocd/utils.go:25]: `StructToMap` or `InterfaceToMap`?

I didn't see any `struct` in this function.

File: pkg/api/gitops/v1alpha1/application.go [hunk 2]
```diff
@@ -25,9 +25,65 @@ type ApplicationSpec struct {
 	ArgoApp *ArgoApplication `json:"argoApp,omitempty"`
 }
 
-// ArgoApplication represents an ArgoCD Application
-// The fields simply are copied from the argo-cd project
+// ArgoApplication is a definition of Argo Application resource.
 type ArgoApplication struct {
+	Spec      ArgoApplicationSpec `json:"spec,omitempty"`
+	Operation *Operation          `json:"operation,omitempty" protobuf:"bytes,4,opt,name=operation"`
```
Comment (LinuxSuRen) [inline pkg/api/gitops/v1alpha1/application.go:31]: do we really need `protobuf`? Please remove it if does not.

File: pkg/api/gitops/v1alpha1/application.go [hunk 3]
```diff
@@ -25,9 +25,65 @@ type ApplicationSpec struct {
...
+	// InitiatedBy contains information about who initiated the operations
+	InitiatedBy OperationInitiator `json:"initiatedBy,omitempty" protobuf:"bytes,2,opt,name=initiatedBy"`
+	// Info is a list of informational items for this operation
+	Info []*Info `json:"info,omitempty" protobuf:"bytes,3,name=info"`
+	// Retry controls the strategy to apply if a sync fails
+	Retry RetryStrategy `json:"retry,omitempty" protobuf:"bytes,4,opt,name=retry"`
+}
+
+// ArgoApplicationSpec represents an ArgoCD Application
+// The fields simply are copied from the argo-cd project
```
Comment (LinuxSuRen) [inline pkg/api/gitops/v1alpha1/application.go:85]: How about putting this statement on `type ArgoApplication struct`? Or put it on `type Operation struct` as well.

**Review by LinuxSuRen — COMMENTED** [review]
Most of the changes are ok to me. But I still left some comments. Please take a look at them.

File: controllers/argocd/application_controller.go [hunk 4]
```diff
@@ -78,70 +77,53 @@ func (r *ApplicationReconciler) reconcileArgoApplication(app *v1alpha1.Applicati
...
+			err = r.Update(ctx, argoApp)
 		}
 	}
 	return
 }
 
 func createUnstructuredApplication(app *v1alpha1.Application) (result *unstructured.Unstructured, err error) {
-	app = app.DeepCopy()
-	argoApp := app.Spec.ArgoApp
+	argoApp := app.Spec.ArgoApp.DeepCopy()
```
[the remark this answers is not in the document]
Comment (JohnNiang) [reply controllers/argocd/application_controller.go:93]: You can see the implementation of ArgoApplication#DeepCopy method:

```go
// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new ArgoApplication.
func (in *ArgoApplication) DeepCopy() *ArgoApplication {
	if in == nil {
		return nil
	}
	out := new(ArgoApplication)
	in.DeepCopyInto(out)
	return out
}
```

The DeepCopy will return nil if the ArgoApplication is nil. But I will change the code to make `app.Spec.ArgoApp != nil` clear.

**Review by LinuxSuRen — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**